src/Uno.UI/Extensions/PermissionsHelper.Android.cs

```diff
@@ -19,11 +19,185 @@ public static void Initialize()
+			// https://developer.android.com/reference/android/content/pm/PackageInstaller.SessionParams.html#setWhitelistedRestrictedPermissions(java.util.Set%3Cjava.lang.String%3E)
+
+			// do we have declared permissions in Manifest?
+			var manifestPermissions = GetManifestPermissions();
+			if (manifestPermissions is null)
+			{
+				return null;
+			}
+
+
+			// test required permissions
+			if (!AreAllPermissionDeclared(manifestPermissions, requiredPermissions))
+			{
+				return null;
+			}
+
+			// prepare list of all permissions (all required, and all optional defined in Manifest
+			var allPermissions = PermissionWeWant(manifestPermissions, requiredPermissions, optionalPermissions);
+
+			// prepare list of permission to ask for
+			var askForPermission = PermissionToAsk(allPermissions);
+
+			return askForPermission.ToArray();
+
+		}
+
+		/// <summary>
+		/// Return true if granted are: all requiredPermissions, and all optionalPermissions mentioned in Manifest. Show dialog asking for missing permissions. Both parameters can be null.
+		/// </summary>
+		/// <param name="requiredPermissions">Permissions that are required. Can be null.</param>
+		/// <param name="optionalPermissions">Permissions that are required only if declared in Manifest. Can be null.</param>
+		/// <param name="ignoreErrors">if 'true', then do not throw exception on error, and simply retur false</param>
+		/// <returns>Bool if all permissions are granted, false if any of them is not granted</returns>
+
+		private static async Task<bool> TryAskPermissionAsync(CancellationToken ct, string[] requiredPermissions, string[] optionalPermissions, bool ignoreErrors)
+		{
+
+			var askForPermission = MissingPermissions(requiredPermissions, optionalPermissions);
+			if (askForPermission is null)
+			{
+				if (ignoreErrors)
+				{
+					return false;
+				}
+
+				throw new InvalidOperationException("MissingPermissions returned ERROR - check Manifest file");
+			}
+
+			if (askForPermission.Count() < 1)
+			{
+				return true;
+			}
+
+			// system dialog asking for permission
+
+			var code = Interlocked.Increment(ref _permissionRequest);
+			var tcs = new TaskCompletionSource<BaseActivity.RequestPermissionsResultWithResultsEventArgs>();
+
+			void handler(object sender, BaseActivity.RequestPermissionsResultWithResultsEventArgs e)
+			{
+
+				if (e.RequestCode == code)
+				{
+					tcs.TrySetResult(e);
+				}
+			}
+
+			var current = BaseActivity.Current;
+
+			try
+			{
+				using (ct.Register(() => tcs.TrySetCanceled()))
+				{
+					current.RequestPermissionsResultWithResults += handler;
 
-		private static int _permissionRequest;
+					ActivityCompat.RequestPermissions(BaseActivity.Current, askForPermission, code);
 
+					var result = await tcs.Task;
+
+					if (result.GrantResults.Length > 0)
+					{
+						bool allGranted = result.GrantResults.All(r => r == Android.Content.PM.Permission.Granted);
+						return allGranted;
+					}
+					return false;   // some error?

```

Comment:
```suggestion
					return false;

```

Reply:
"some error" means: this is some error condition, as we should never get into this `return` (`result.GrantResults.Length` should always be > 0), but we don't throw exception here. Maybe we should...